spinalcordtoolbox/types.py

```diff
@@ -534,19 +534,18 @@ def compute_vertebral_distribution(self, discs_levels, label_reference='C1'):
 
     def get_closest_to_relative_position(self, vertebral_level, relative_position):
         """
-        :param vertebral_level:
+        :param vertebral_level: the name of a vertebral level, as a string
         :param relative_position: the relative position [0, 1] from upper disc
         """
-        indexes_vert = np.argwhere(np.array(self.l_points) == vertebral_level)
-        if len(indexes_vert) == 0:
+        dist_indices = [
+            (abs(self.dist_points_rel[i] - relative_position), i)
+            for i in range(self.number_of_points)
+            if self.l_points[i] == vertebral_level
```

Comment:
I'm also unhappy about `self.number_of_points`. Most of the time I'd rather iterate over the relevant list, possibly with a `zip` or an `enumerate`, and avoid using `range(self.number_of_points)` entirely.

But the difficulty here is that `self.number_of_points` is the length of _so many_ different lists:
```py3
self.points
self.coordinate_systems
self.plans_parameters
self.matrices
self.inverse_matrices
self.offset_plans
self.progressive_length
self.incremental_length
self.progressive_length_inverse
self.incremental_length_inverse
self.l_points
self.dist_points
self.dist_points_rel
```
(Not counting local variables.)

For `self.l_points` specifically, I'm not too worried: both `self.points` and `self.number_of_points` seem to be used in a read-only manner, and `self.l_points` is only ever created to be a list of length `self.number_of_points`. It's initialized with dummy values that get replaced in a loop, rather than created by appending values to an empty list, as you might otherwise expect.

(I'm not very happy that the dummy value for `self.l_points` is the integer 0, when the normal values are strings, but that's a different discussion.)

To me, having so many lists of the same length means that maybe the code could be better structured by having a single list of instances of a (hypothetical) `CenterlinePoint` class. But I ran out of refactoring steam before getting there.

Reply:
> For `self.l_points` specifically, I'm not too worried: both `self.points` and `self.number_of_points` seem to be used in a read-only manner, and `self.l_points` is only ever created to be a list of length `self.number_of_points`.

Perfect! That satisfies my main worry as well.